staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go

```diff
@@ -70,32 +73,34 @@ func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.Endpo
 		return nil, err
 	}
 
-	eps, err := endpoints.Endpoints(namespace).Get(svc.Name)
+	endpointSliceSelector := labels.SelectorFromSet(labels.Set{discoveryv1.LabelServiceName: svc.Name})
+	slices, err := endpointSlices.EndpointSlices(namespace).List(endpointSliceSelector)
 	if err != nil {
 		return nil, err
 	}
-	if len(eps.Subsets) == 0 {
+	if len(slices) == 0 {
 		return nil, errors.NewServiceUnavailable(fmt.Sprintf("no endpoints available for service %q", svc.Name))
 	}
 
-	// Pick a random Subset to start searching from.
-	ssSeed := rand.Intn(len(eps.Subsets))
+	// Find a slice that has the port. (Of course there is probably only 1 slice.)
+	for _, slice := range slices {
+		for i := range slice.Ports {
+			if slice.Ports[i].Name == nil || *slice.Ports[i].Name != svcPort.Name {
+				continue
+			}
 
-	// Find a Subset that has the port.
-	for ssi := 0; ssi < len(eps.Subsets); ssi++ {
-		ss := &eps.Subsets[(ssSeed+ssi)%len(eps.Subsets)]
-		if len(ss.Addresses) == 0 {
-			continue
-		}
-		for i := range ss.Ports {
-			if ss.Ports[i].Name == svcPort.Name {
-				// Pick a random address.
-				ip := ss.Addresses[rand.Intn(len(ss.Addresses))].IP
-				port := int(ss.Ports[i].Port)
-				return &url.URL{
-					Scheme: "https",
-					Host:   net.JoinHostPort(ip, strconv.Itoa(port)),
-				}, nil
+			// Starting from a random index, find a Ready endpoint
+			offset := rand.Intn(len(slice.Endpoints))
+			for epi := range slice.Endpoints {
+				ep := &slice.Endpoints[(epi+offset)%len(slice.Endpoints)]
+				if ep.Conditions.Ready == nil || *ep.Conditions.Ready {
+					ip := ep.Addresses[0]
+					port := int(*slice.Ports[i].Port)
```

Comment:
godoc on the Port field sounds like Port can be nil... should we check / require a non-nil port value above when deciding if the Port is valid for our use?

Reply:
Huh. This is another unused future-extensibility point. Port is allowed to be unset (and thus not validated as required) even though it's required to be set for EndpointSlices corresponding to Services. Grumble.

Reply:
Yeah... I just don't know the EndpointSlice API surface well enough to know which of these "allowed by the spec" things we have to handle, so my default is to check everything we could possibly encounter